reactivesocket-core/src/main/java/io/reactivesocket/frame/FrameHeaderFlyweight.java

```diff
@@ -219,40 +229,77 @@ public static ByteBuffer sliceFrameMetadata(final DirectBuffer directBuffer, fin
         return result;
     }
 
-    private static int frameLength(final DirectBuffer directBuffer, final int offset, final int externalFrameLength) {
-        int frameLength = externalFrameLength;
-
-        if (INCLUDE_FRAME_LENGTH) {
-            frameLength = directBuffer.getInt(offset + FRAME_LENGTH_FIELD_OFFSET, ByteOrder.BIG_ENDIAN);
+    static int frameLength(final DirectBuffer directBuffer, final int offset, final int externalFrameLength) {
+        if (!INCLUDE_FRAME_LENGTH) {
+            return externalFrameLength;
         }
 
-        return frameLength;
+        return decodeLength(directBuffer, offset + FRAME_LENGTH_FIELD_OFFSET);
     }
 
-    private static int computeMetadataLength(final int metadataPayloadLength) {
-        return metadataPayloadLength + (0 == metadataPayloadLength? 0 : BitUtil.SIZE_OF_INT);
+    private static int metadataFieldLength(final DirectBuffer directBuffer, final int offset) {
+        return computeMetadataLength(metadataLength(directBuffer, offset));
     }
 
-    private static int metadataFieldLength(final DirectBuffer directBuffer, final int offset) {
+    private static int metadataLength(final DirectBuffer directBuffer, final int offset) {
+        return metadataLength(directBuffer, offset, metadataOffset(directBuffer, offset));
+    }
+
+    static int metadataLength(final DirectBuffer directBuffer, final int offset, final int metadataOffset) {
         int metadataLength = 0;
 
-        short flags = directBuffer.getShort(offset + FLAGS_FIELD_OFFSET, ByteOrder.BIG_ENDIAN);
+        int flags = flags(directBuffer, offset);
         if (FLAGS_M == (FLAGS_M & flags)) {
-            metadataLength = directBuffer.getInt(metadataOffset(directBuffer, offset), ByteOrder.BIG_ENDIAN) & 0xFFFFFF;
+            metadataLength = decodeLength(directBuffer, metadataOffset);
         }
 
         return metadataLength;
     }
 
+    private static int computeMetadataLength(final int length) {
+        return length == 0 ? 0 : length + FRAME_LENGTH_SIZE;
+    }
+
+    private static void encodeLength(
+            final MutableDirectBuffer mutableDirectBuffer,
+            final int offset,
+            final int length
+    ) {
+        if ((length & ~FRAME_LENGTH_MASK) != 0) {
+            throw new IllegalArgumentException("Length is larger than 24 bits");
```

Comment:
Yes, the error message could say it's either larger, or negative, since a negative number would fail this test (as it should).